File: common/src/clock.c
#include "lpc111x_min.h"
#include "clock.h"

void clock_init_48mhz(void)
{
    /* Power up SYSPLL and ensure IRC is powered. Reserved bits per UM10398 Table 44. */
    uint32_t pdruncfg = LPC_SYSCON_PDRUNCFG;
    pdruncfg &= ~((1u << 7) | (1u << 1) | (1u << 0));
    pdruncfg |= (1u << 8) | (1u << 10) | (1u << 11) | (7u << 13);
    pdruncfg &= ~((1u << 9) | (1u << 12));
    LPC_SYSCON_PDRUNCFG = pdruncfg;

    /* Select IRC as PLL source. */
    LPC_SYSCON_SYSPLLCLKSEL = 0x0u;
    LPC_SYSCON_SYSPLLCLKUEN = 0x0u;
    LPC_SYSCON_SYSPLLCLKUEN = 0x1u;

    /* Configure PLL: M = 4 (MSEL=3), P = 2 (PSEL=1). */
    LPC_SYSCON_SYSPLLCTRL = (3u << 0) | (1u << 5);

    /* Wait for PLL lock. */
    while ((LPC_SYSCON_SYSPLLSTAT & 0x1u) == 0u) {
        /* spin */
    }

    /* Set main clock to SYSPLL clock out. */
    LPC_SYSCON_MAINCLKSEL = 0x3u;
    LPC_SYSCON_MAINCLKUEN = 0x0u;
    LPC_SYSCON_MAINCLKUEN = 0x1u;

    /* AHB divider = 1 (system clock = main clock). */
    LPC_SYSCON_SYSAHBCLKDIV = 1u;
}

File: common/rust/src/clock.rs
#![allow(dead_code)]

use crate::pac::Peripherals;

pub fn init_48mhz() {
    let p = unsafe { Peripherals::steal() };

    // Power up SYSPLL and ensure IRC is powered; keep reserved bits as in UM10398 Table 44.
    let mut pdruncfg = p.SYSCON.pdruncfg.read().bits();
    pdruncfg &= !((1 << 7) | (1 << 1) | (1 << 0));
    pdruncfg |= (1 << 8) | (1 << 10) | (1 << 11) | (7 << 13);
    pdruncfg &= !((1 << 9) | (1 << 12));
    unsafe { p.SYSCON.pdruncfg.write(|w| w.bits(pdruncfg)) };

    // Select IRC as PLL source and update.
    unsafe {
        p.SYSCON.syspllclksel.write(|w| w.bits(0));
        p.SYSCON.syspllclkuen.write(|w| w.bits(0));
        p.SYSCON.syspllclkuen.write(|w| w.bits(1));
    }
    while (p.SYSCON.syspllclkuen.read().bits() & 0x1) == 0 {}

    // Configure PLL: M = 4 (MSEL=3), P = 2 (PSEL=1).
    unsafe { p.SYSCON.syspllctrl.write(|w| w.bits((3 << 0) | (1 << 5))) };

    // Wait for PLL lock.
    while p.SYSCON.syspllstat.read().bits() & 0x1 == 0 {}

    // Main clock = SYSPLL clock out.
    unsafe {
        p.SYSCON.mainclksel.write(|w| w.bits(3));
        p.SYSCON.mainclkuen.write(|w| w.bits(0));
        p.SYSCON.mainclkuen.write(|w| w.bits(1));
        p.SYSCON.sysahbclkdiv.write(|w| w.bits(1));
    }
    while (p.SYSCON.mainclkuen.read().bits() & 0x1) == 0 {}
}
